art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py

```diff
@@ -0,0 +1,447 @@
+# MIT License
+#
+# Copyright (C) IBM Corporation 2020
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+# persons to whom the Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements the Projected Gradient Descent attack `ProjectedGradientDescent` as an iterative method in which,
+after each iteration, the perturbation is projected on an lp-ball of specified radius (in addition to clipping the
+values of the adversarial sample so that it lies in the permitted data range). This is the attack proposed by Madry et
+al. for adversarial training.
+
+| Paper link: https://arxiv.org/abs/1706.06083
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+
+import numpy as np
+import tensorflow as tf
+from scipy.stats import truncnorm
+
+from art.config import ART_NUMPY_DTYPE
+from art.estimators.estimator import BaseEstimator, LossGradientsMixin
+from art.attacks.attack import EvasionAttack
+from art.utils import compute_success, get_labels_np_array, check_and_transform_label_format, random_sphere
+
+logger = logging.getLogger(__name__)
+
+
+class ProjectedGradientDescentTensorflowV2(EvasionAttack):
```

Comment:
See comment about parent class for `projected_gradient_descent_pytorch.py`.

Reply:
No, see explain above.

Reply:
Yes, this is done with a new parent class for all implementation of PGD.